### DISCUSSION TITLE
Found advantages of crafting emacs.d without use-package and implementation

### DISCUSSION THREAD

**Discussion by MorphicResonance** [SystemCrafters/crafted-emacs#385] [Ideas]
The current project is notable for the fact that it does not apply the use-package configurator, but strives assemble sequence of packages and settings from a clean
Elisp. 
Despite the apparent complexity of configuration settings, this provides additional benefits. 

Although now use-package is built into
emacs and it makes sense to explore it when managing several dozens of packages, it dictates us its own rules, which we will have to obey.

Having looked at some old school emacs.d files from emacs.nz, which were crafted without use-package, found some interesting points that I wanted would implement here. The [kiss](https://github.com/bling/dotemacs) configuration is especially noteworthy..
- `with-eval-after-load` _lets you defer execution of code until after a feature has been loaded_. If this is combined with modularity of crafted-emacs.., is it possible to create such thing that only requested modules will be loaded 'on demand'?
- _bindings in one place_. Is it possible to achieve this here so that user is able to collect all keyboard shortcuts in one file?
I recommend paying attention to this dotemacs.

Another fully functional [configuration of a new school without use-package](https://gitlab.com/aimebertrand/dotemacs/) by Aimé Bertrand. Conspicuous ordering of custom variables and strict adherence to rules.

**Comment by jeffbowman** [comment]
We use `with-eval-after-load` as well in several modules.  To answer your question about loading only requested modules on demand: yes, of course.  There are some caveats:

* Modules/packages which do not provide an `autoload` function that would naturally be called (for example a minor or major mode, or some command bound to a keybinding) would need to have one provided so that it gets loaded when it is demanded.  See the `autoload` function for how to do that.
* Some variables may need to be provided through a `defvar` declaration to avoid compilation errors, since they may not be available until after the module/package is loaded. 

Bindings in one place is as simple as putting all of your bindings in a single file and then loading that file.  If you are using Crafted Emacs, and have a `custom-modules` folder, you could put a `my-bindings.el` file there which contains only keybindings.  You would then simply `(require 'my-bindings)` in  your `init.el` file where appropriate, probably after the configuration phase.  If you aren't using Crafted Emacs, you can still follow this advice, but you'll need either manually add the path to where the `my-bindings.el` file lives to your load path, or use `load` with an explicit path to the file.
